src/gf.c

```diff
@@ -177,53 +223,187 @@ void jl_mk_builtin_func(jl_datatype_t *dt, const char *name, jl_fptr_t fptr)
-jl_code_info_t *jl_type_infer(jl_method_instance_t *li, int force)
+jl_code_info_t *jl_type_infer(jl_method_instance_t **pli, size_t world, int force)
 {
     JL_TIMING(INFERENCE);
     if (jl_typeinf_func == NULL)
         return NULL;
     jl_code_info_t *src = NULL;
 #ifdef ENABLE_INFERENCE
+    jl_method_instance_t *li = *pli;
     jl_module_t *mod = NULL;
     if (li->def != NULL)
         mod = li->def->module;
     static int inInference = 0;
     int lastIn = inInference;
+    size_t last_age = jl_get_ptls_states()->world_age;
     inInference = 1;
     if (force ||
-        (mod != jl_gf_mtable(jl_typeinf_func)->module &&
+        (last_age != jl_typeinf_world &&
+         mod != jl_gf_mtable(jl_typeinf_func)->module &&
          (mod != jl_core_module || !lastIn))) { // avoid any potential recursion in calling jl_typeinf_func on itself
-        assert(li->inInference == 0);
-        jl_value_t *fargs[2];
+        assert(li->inInference == 0 && "unexpectedly asked to infer a method that is already being inferred");
+        jl_value_t **fargs;
+        JL_GC_PUSHARGS(fargs, 3);
         fargs[0] = (jl_value_t*)jl_typeinf_func;
         fargs[1] = (jl_value_t*)li;
+        fargs[2] = jl_box_ulong(world);
 #ifdef TRACE_INFERENCE
         jl_printf(JL_STDERR,"inference on ");
         jl_static_show_func_sig(JL_STDERR, (jl_value_t*)li->specTypes);
         jl_printf(JL_STDERR, "\n");
 #endif
-        src = (jl_code_info_t *)jl_apply(fargs, 2);
-        if (src == (void*)jl_nothing)
-            src = NULL;
-        assert(li->def || li->inInference == 0); // if this is toplevel expr, make sure inference finished
+        jl_get_ptls_states()->world_age = jl_typeinf_world;
+        jl_svec_t *linfo_src_rettype = (jl_svec_t*)jl_apply(fargs, 3);
+        jl_get_ptls_states()->world_age = last_age;
+        assert((li->def || li->inInference == 0) && "inference failed on a toplevel expr");
+        if (jl_is_svec(linfo_src_rettype) && jl_svec_len(linfo_src_rettype) == 3 &&
+            jl_is_method_instance(jl_svecref(linfo_src_rettype, 0)) &&
+            jl_is_code_info(jl_svecref(linfo_src_rettype, 1))) {
+            *pli = (jl_method_instance_t*)jl_svecref(linfo_src_rettype, 0);
+            src = (jl_code_info_t*)jl_svecref(linfo_src_rettype, 1);
+        }
+        JL_GC_POP();
     }
     inInference = lastIn;
 #endif
     return src;
 }
 
-JL_DLLEXPORT void jl_set_lambda_rettype(jl_method_instance_t *li, jl_value_t *rettype, int32_t const_flags, jl_value_t *inferred_const, jl_value_t *inferred)
+
+static int jl_is_rettype_inferred(jl_method_instance_t *li)
+{
+    if (!li->inferred)
+        return 0;
+    if (jl_is_code_info(li->inferred) && !((jl_code_info_t*)li->inferred)->inferred)
+        return 0;
+    return 1;
+}
+
+
+struct set_world {
+    jl_method_instance_t *replaced;
+    size_t world;
+};
+static int set_max_world2(jl_typemap_entry_t *entry, void *closure0)
```

Comment:
no, usually it's redundant with "set". In this case, I thought there might be multiple useful ways that a `set_world` function might manage this update which were safer. In the PR, this one turned out to be the only valid one. But I'll reintroduce others as needed in post-PR fixes.